File: gatling-recorder/src/main/java/com/excilys/ebi/gatling/recorder/configuration/Configuration.java
/**
 * Copyright 2011-2012 eBusiness Information, Groupe Excilys (www.excilys.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 		http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.excilys.ebi.gatling.recorder.configuration;

import static org.apache.commons.lang.StringUtils.EMPTY;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import com.excilys.ebi.gatling.recorder.ui.enumeration.FilterStrategy;

public class Configuration {

	public static Configuration getInstance() {
		return instance;
	}

	public static void initFromExistingConfig(Configuration c) {
		instance.setPort(c.getPort());
		instance.setSslPort(c.getSslPort());
		instance.setProxy(c.getProxy());
		instance.setFollowRedirect(c.isFollowRedirect());
		instance.setSimulationClassName(c.getSimulationClassName());
		instance.setSimulationPackage(c.getSimulationPackage());
		instance.setFilterStrategy(c.getFilterStrategy());
		instance.setPatterns(c.getPatterns());
		instance.setOutputFolder(c.getOutputFolder());
		instance.setSaveConfiguration(true);
		instance.setEncoding(c.getEncoding());
	}

	public static void initFromCommandLineOptions(CommandLineOptions c) {
		instance.setPort(c.getLocalPort());
		instance.setSslPort(c.getLocalPortSsl());
		instance.getProxy().setHost(c.getProxyHost());
		instance.getProxy().setPort(c.getProxyPort());
		instance.getProxy().setSslPort(c.getProxyPortSsl());
		if (c.getOutputFolder() != null)
			instance.setOutputFolder(c.getOutputFolder());

		if (c.getSimulationClassName() != null)
			instance.setSimulationClassName(c.getSimulationClassName());
		instance.setSimulationPackage(c.getSimulationPackage());
		instance.setRequestBodiesFolder(c.getRequestBodiesFolder());
		if (c.getEncoding() != null)
			instance.setEncoding(c.getEncoding());
	}

	private static final Configuration instance = new Configuration();

	private int port = 8000;
	private int sslPort = 8001;
	private ProxyConfig proxy = new ProxyConfig();
	private boolean followRedirect;
	private FilterStrategy filterStrategy = FilterStrategy.NONE;
	private List<Pattern> patterns = new ArrayList<Pattern>();
	private String outputFolder = System.getenv("GATLING_HOME") + File.separator + "user-files" + File.separator + "simulations";
	private boolean saveConfiguration;
	private String encoding = "UTF-8";
	private String requestBodiesFolder = System.getenv("GATLING_HOME") + File.separator + "user-files" + File.separator + "requests-bodies";
	private String simulationPackage = EMPTY;
	private String simulationClassName = "Simulation";

	private Configuration() {

	}

	@Override
	public String toString() {
		return new StringBuilder().append("Configuration [port=").append(port).append(", sslPort=").append(sslPort).append(", proxy=").append(proxy).append(", filterStrategy=")
				.append(filterStrategy).append(", patterns=").append(patterns).append(", outputFolder=").append(outputFolder).append(", saveConfiguration=")
				.append(saveConfiguration).append("]").toString();
	}

	public int getPort() {
		return port;
	}

	public void setPort(int port) {
		this.port = port;
	}

	public ProxyConfig getProxy() {
		return proxy;
	}

	public void setProxy(ProxyConfig proxy) {
		this.proxy = proxy;
	}

	public FilterStrategy getFilterStrategy() {
		return filterStrategy;
	}

	public void setFilterStrategy(FilterStrategy filterStrategy) {
		this.filterStrategy = filterStrategy;
	}

	public List<Pattern> getPatterns() {
		return patterns;
	}

	public void setPatterns(List<Pattern> patterns) {
		this.patterns = patterns;
	}

	public String getOutputFolder() {
		return outputFolder;
	}

	public void setOutputFolder(String outputFolder) {
		this.outputFolder = outputFolder;
	}

	public int getSslPort() {
		return sslPort;
	}

	public void setSslPort(int sslPort) {
		this.sslPort = sslPort;
	}

	public boolean isSaveConfiguration() {
		return saveConfiguration;
	}

	public void setSaveConfiguration(boolean saveConfiguration) {
		this.saveConfiguration = saveConfiguration;
	}

	public String getSimulationPackage() {
		return simulationPackage;
	}

	public void setSimulationPackage(String simulationPackage) {
		this.simulationPackage = simulationPackage;
	}

	public String getSimulationClassName() {
		return simulationClassName;
	}

	public void setSimulationClassName(String simulationClassName) {
		this.simulationClassName = simulationClassName;
	}

	public String getRequestBodiesFolder() {
		return requestBodiesFolder;
	}

	public void setRequestBodiesFolder(String requestBodiesFolder) {
		this.requestBodiesFolder = requestBodiesFolder;
	}

	public String getEncoding() {
		return encoding;
	}

	public void setEncoding(String encoding) {
		this.encoding = encoding;
	}

	public boolean isFollowRedirect() {
		return followRedirect;
	}

	public void setFollowRedirect(boolean followRedirect) {
		this.followRedirect = followRedirect;
	}
}

File: gatling-recorder/src/main/scala/com/excilys/ebi/gatling/recorder/config/Configuration.scala
/**
 * Copyright 2011-2012 eBusiness Information, Groupe Excilys (www.excilys.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 		http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.excilys.ebi.gatling.recorder.config

import java.io.{ IOException, FileWriter }

import scala.reflect.BeanProperty
import scala.tools.nsc.io.Path.string2path
import scala.tools.nsc.io.File

import com.excilys.ebi.gatling.core.config.{ GatlingFiles, GatlingConfiguration }
import com.excilys.ebi.gatling.core.util.IOHelper.use
import com.excilys.ebi.gatling.recorder.ui.Commons.GATLING_RECORDER_FILE_NAME
import com.excilys.ebi.gatling.recorder.ui.enumeration.FilterStrategy.NONE
import com.thoughtworks.xstream.io.xml.DomDriver
import com.thoughtworks.xstream.XStream

import grizzled.slf4j.Logging

object Configuration extends Logging {

	GatlingConfiguration.setUp(None, None, None, None, None)
	val DEFAULT_CLASS_NAME = "Simulation"

	private val XSTREAM = {
		val xstream = new XStream(new DomDriver)
		xstream.alias("configuration", classOf[Configuration])
		xstream.alias("pattern", classOf[Pattern])
		xstream.alias("proxy", classOf[ProxyConfig])
		xstream
	}
	private val CONFIGURATION_FILE = File(System.getProperty("user.home") / GATLING_RECORDER_FILE_NAME)

	val configuration = new Configuration

	def apply(options: Options) {
		initFromDisk
		initFromCli(options)
	}

	def apply(c: Configuration) {
		configuration.port = c.port
		configuration.sslPort = c.sslPort
		configuration.proxy = c.proxy
		configuration.filterStrategy = c.filterStrategy
		configuration.patterns = c.patterns
		configuration.outputFolder = c.outputFolder
		configuration.simulationClassName = c.simulationClassName
		configuration.simulationPackage = c.simulationPackage
		configuration.followRedirect = c.followRedirect
		configuration.automaticReferer = c.automaticReferer
		configuration.saveConfiguration = true
		configuration.encoding = c.encoding
	}

	def saveToDisk {
		use(new FileWriter(CONFIGURATION_FILE.jfile)) { fw =>
			try {
				XSTREAM.toXML(configuration, fw)
			} catch {
				case e: IOException => error(e.getMessage)
			}
		}
	}

	private def initFromDisk {
		if (CONFIGURATION_FILE.exists) {
			try {
				val c = XSTREAM.fromXML(CONFIGURATION_FILE.jfile).asInstanceOf[Configuration]
				Configuration(c)
			} catch {
				case e: Exception => error(e.getMessage)
			}
		}
	}

	private def initFromCli(o: Options) {
		o.localPort.map(configuration.port = _)
		o.localPortSsl.map(configuration.sslPort = _)
		for {
			val proxyHost <- o.proxyHost
			val proxyPort <- o.proxyPort
		} {
			configuration.proxy = new ProxyConfig(Some(proxyHost), Some(proxyPort), o.proxyPortSsl, None, None)
		}
		o.outputFolder.map(configuration.outputFolder = _)
		o.simulationClassName.map(configuration.simulationClassName = _)
		o.simulationPackage.map(pkg => configuration.simulationPackage = Some(pkg))
		o.requestBodiesFolder.map(configuration.requestBodiesFolder = _)
		o.encoding.map(configuration.encoding = _)
		o.followRedirect.map(configuration.followRedirect = _)
		o.automaticReferer.map(configuration.automaticReferer = _)
	}
}

class Configuration {

	@BeanProperty var port = 8000
	@BeanProperty var sslPort = 8001
	@BeanProperty var proxy = new ProxyConfig
	@BeanProperty var filterStrategy = NONE
	@BeanProperty var patterns: List[Pattern] = Nil
	@BeanProperty var outputFolder: String = Option(System.getenv("GATLING_HOME")).map(_ => GatlingFiles.simulationsFolder.toString).getOrElse(System.getProperty("user.home"))
	@transient var saveConfiguration = false
	@BeanProperty var encoding = "UTF-8"
	@transient var requestBodiesFolder: String = GatlingFiles.requestBodiesFolder.toString
	@BeanProperty var simulationPackage: Option[String] = None
	@BeanProperty var simulationClassName: String = Configuration.DEFAULT_CLASS_NAME
	@BeanProperty var followRedirect: Boolean = true
	@BeanProperty var automaticReferer: Boolean = true

	override def toString =
		new StringBuilder("Configuration [")
			.append("port=").append(port).append(", ")
			.append("sslPort=").append(sslPort).append(", ")
			.append("proxy=").append(proxy).append(", ")
			.append("filterStrategy=").append(filterStrategy).append(", ")
			.append("patterns=").append(patterns).append(", ")
			.append("outputFolder=").append(outputFolder).append(", ")
			.append("saveConfiguration=").append(saveConfiguration).append(", ")
			.append("encoding=").append(encoding).append(", ")
			.append("requestBodiesFolder=").append(requestBodiesFolder).append(", ")
			.append("simulationPackage=").append(simulationPackage).append(", ")
			.append("simulationClassName=").append(simulationClassName).append(", ")
			.append("followRedirect=").append(followRedirect)
			.append("automaticReferer=").append(automaticReferer)
			.append("]").toString
}
